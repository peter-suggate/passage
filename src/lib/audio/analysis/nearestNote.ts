import { Observable } from "rxjs";
import { map as map$, filter as filter$ } from "rxjs/operators";
import { AudioRecorderEventTypes } from "../recorder";
import { AudioPitchEvent } from "../audio-types";
import {
  Note,
  NOTES,
  CONCERT_PITCH,
  CONCERT_A_HALFTONE_INDEX,
} from "@/lib/scales";

export type NearestNote = {
  value: Note;
  octave: number;
  cents: number;
};

export const frequencyToNearestNote = (
  hz: number,
  concertPitch = CONCERT_PITCH
): NearestNote => {
  const nearestNote = Math.round(12 * Math.log2(hz / concertPitch));
  const noteAndOctave = CONCERT_A_HALFTONE_INDEX + nearestNote;
  const nearestNoteHz = concertPitch * Math.pow(2, nearestNote / 12);

  const result = {
    value: NOTES[noteAndOctave % 12],
    octave: Math.round(noteAndOctave / 12) | 0,
    cents: 1200 * Math.log2(hz / nearestNoteHz),
  };

  // console.log("result", result, "hz", hz, "nearestNoteHz", nearestNoteHz);
  return result;
};

export const noteToFrequency = (semitonesFromConcertA: number) => {
  return CONCERT_PITCH * Math.pow(2, semitonesFromConcertA / 12);
};

const isPitchEvent = (e: AudioRecorderEventTypes): e is AudioPitchEvent =>
  e.type === "pitch";

export const activeNote$ = (
  pitches$: Observable<AudioRecorderEventTypes>
): Observable<NearestNote> => {
  return pitches$.pipe(
    filter$<AudioRecorderEventTypes, AudioPitchEvent>(isPitchEvent),
    map$((e) => frequencyToNearestNote(e.pitch.frequency))
  );
};
