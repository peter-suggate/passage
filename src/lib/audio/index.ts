import { AudioService, makeAudioService, makeAudio$ } from "./audioService";

export const audioService: AudioService = makeAudioService();

export const audio$ = makeAudio$(audioService);
