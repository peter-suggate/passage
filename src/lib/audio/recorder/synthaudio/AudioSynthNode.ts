import { Subject } from "rxjs";
import {
  AudioRecorderEventTypes,
  AudioProcessorEventTypes,
} from "../recorder-types";
import init, {
  AudioSamplesProcessor,
  PitchDetector,
} from "../../../../../public/music-analyzer-wasm-rs/music_analyzer_wasm_rs.js";
/* eslint-disable @typescript-eslint/no-var-requires */
// const asd = require("../../../../../public/worker-polyfills/text-encoder");
// const wasm = require("../../../../../public/music-analyzer-wasm-rs/music_analyzer_wasm_rs");
// var init, { AudioSamplesProcessor, PitchDetector } = require("../../../../../public/music-analyzer-wasm-rs/music_analyzer_wasm_rs");
// import init, {
//   PitchDetector,
//   AudioSamplesProcessor,
// } from "music-analyzer-wasm-rs/music_analyzer_wasm_rs";

export class AudioSynthNode extends Subject<AudioRecorderEventTypes> {
  private constructor(
    private wasmSamplesProcessor: any /*AudioSamplesProcessor*/,
    private pitchDetector: any /*PitchDetector*/
  ) {
    super();
  }

  private async loadWasm(wasmBytes: ArrayBuffer) {
    await init(WebAssembly.compile(wasmBytes));
  }

  static async create() {
    try {
      console.log("synthesizer: loading wasm");

      // const wasmBytes = await AudioSynthNode.fetchMusicAnalyzerWasm();
      // const wasm = await init(WebAssembly.compile(wasmBytes));

      const wasmSamplesProcessor = AudioSamplesProcessor.new();

      const pitchDetector = wasmSamplesProcessor.create_pitch_detector(
        "McLeod",
        2048
      );

      if (!pitchDetector) {
        throw Error("Wasm pitch detector could not be created");
      }

      return new AudioSynthNode(wasmSamplesProcessor, pitchDetector);
    } catch (e) {
      throw new Error(
        `Failed to load audio analyzer WASM module. Further info: ${e.message}`
      );
    }
  }

  // private static async fetchMusicAnalyzerWasm() {
  //   const res = await globalThis.fetch(
  //     scriptUrl("music-analyzer-wasm-rs/music_analyzer_wasm_rs_bg.wasm")
  //   );
  //   return res.arrayBuffer();
  // }

  onmessage(eventData: AudioProcessorEventTypes) {
    switch (eventData.type) {
      case "initialized": {
        break;
      }
      case "pitches": {
        eventData.result.forEach((pitch) => {
          if (pitch.is_onset) {
            this.next({
              type: "onset",
              t: pitch.t,
            });
          }

          this.next({
            type: "pitch",
            pitch,
          });
        });

        break;
      }
      default:
        break;
    }
  }
}
