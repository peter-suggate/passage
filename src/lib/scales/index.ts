export * from "./scales-types";
export * from "./western-scales";
