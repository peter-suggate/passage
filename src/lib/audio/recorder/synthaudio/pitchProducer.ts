import { NonNegInteger, nonNegInteger, Bpm, bpm, Note } from "@/lib/scales";
import { Piece } from "@/lib/music-recognition";
import { noteToFrequency } from "../../analysis";
import { halftonesFromConcertA } from "@/lib/scales/generateScaleHalftones";

const bpmToMsPerBeat = (bpm: Bpm) => {
  return (1000 * 60) / bpm;
};

export type PitchProducerConfig = {
  piece: Piece;
  startTime: NonNegInteger;
  startNote: Note;
  bpm?: Bpm;
  startOctave?: NonNegInteger;
};

export const pitchProducer = (config: PitchProducerConfig) => (
  ms: NonNegInteger
): number => {
  const { piece, startTime, startNote, startOctave } = config;

  const timeFromStart = nonNegInteger(ms - startTime);

  // Play piece on loop.
  const beatsPerMin = config.bpm || bpm(60);
  const noteIndex =
    Math.floor(timeFromStart / bpmToMsPerBeat(beatsPerMin)) %
    piece.notes.length;

  // The piece's notes are represented as sequence of deltas (of semitones).
  const noteValue = piece.notes
    .slice(0, noteIndex + 1)
    .reduce((acc, cur) => acc + cur, 0);

  return noteToFrequency(
    noteValue +
      halftonesFromConcertA(startNote, startOctave || nonNegInteger(5))
  );
};
