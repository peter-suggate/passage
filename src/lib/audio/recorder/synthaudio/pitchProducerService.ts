/**
 * State machine to help with generating sequences of pitches to help test certain
 * scenarios. e.g. Allow switching from one piece to another or adding regions of
 * silence.
 */

import { createMachine, assign, interpret } from "xstate";
import { Piece } from "@/lib/music-recognition";
import { NonNegInteger } from "@/lib/scales";
import { pitchProducer, PitchProducerConfig } from "./pitchProducer";

export type PitchProducerContext = {
  current: "silence" | Piece;
  producer?: (ms: NonNegInteger) => number;
};

type Event =
  | {
      type: "START_PIECE";
      config: PitchProducerConfig;
    }
  | { type: "PAUSE" };

export type PitchProducerState = {
  context: PitchProducerContext;
} & ({ value: "silence" } | { value: "playing"; piece: Piece });

const pitchProducerMachine = createMachine<
  PitchProducerContext,
  Event,
  PitchProducerState
>(
  {
    id: "PitchProducer",
    initial: "silence",
    context: { current: "silence" },
    states: {
      silence: {
        on: {
          START_PIECE: {
            actions: assign({
              current: (_, e) => e.config.piece,
              producer: (_, e) => pitchProducer(e.config),
            }),
          },
        },
      },
    },
  },
  {
    services: {},
    actions: {},
  }
);

export const makePitchProducerService = () =>
  interpret(pitchProducerMachine)
    // .onTransition((state) => console.log(state.value, state.context))
    .start();

export type PitchProducerService = ReturnType<typeof makePitchProducerService>;
