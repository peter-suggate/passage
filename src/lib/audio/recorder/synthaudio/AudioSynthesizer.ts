import { Subject, of, Subscription } from "rxjs";
import { AudioRecorderEventTypes } from "../recorder-types";
import init, {
  AudioSamplesProcessor,
  PitchDetector,
} from "../../../../../public/music-analyzer-wasm-rs/music_analyzer_wasm_rs";
import { SynthesizerConfig } from "../../synth/synth-types";
import { animationFrame } from "rxjs/internal/scheduler/animationFrame";
import { repeat, map } from "rxjs/operators";
import { Pitch } from "music-analyzer-wasm-rs";
import { makePitchProducerService } from "./pitchProducerService";
import { nonNegInteger } from "@/lib/scales";
import { PitchProducerConfig } from "./PitchProducer";

function isTest() {
  return process.env.JEST_WORKER_ID !== undefined;
}

const SinewaveGenerator = (sampleRate: number) => {
  let x = 0;

  const TWO_PI = Math.PI * 2;
  const invSampleRate = 1 / sampleRate;

  return {
    next: (atFreq: number) => {
      x += TWO_PI * atFreq * invSampleRate;
      return Math.sin(x);
    },
  };
};

export interface PieceSynthesizer {
  /**
   * Stop whatever we're currently playing and start producing ntoes for the given piece.
   *
   * @param piece A description of the piece to switch to.
   */
  switchTo(config: PitchProducerConfig): void;
}

export class AudioSynthesizer extends Subject<AudioRecorderEventTypes>
  implements PieceSynthesizer {
  private memo = new Map();

  static SAMPLE_RATE = 48000;
  static CHUNK_SIZE = 128;
  static UPDATES_PER_SECOND =
    AudioSynthesizer.SAMPLE_RATE / AudioSynthesizer.CHUNK_SIZE;

  prevNoteFreq = 0;
  prevNoteT = 0;
  wasmChunkBuffer = new Float32Array(AudioSynthesizer.CHUNK_SIZE);
  sinewaveGenerator = SinewaveGenerator(AudioSynthesizer.SAMPLE_RATE);

  private constructor(
    private readonly config: SynthesizerConfig,
    private readonly wasmSamplesProcessor: AudioSamplesProcessor,
    private readonly pitchDetector: PitchDetector,
    private readonly pitchProducer = makePitchProducerService()
  ) {
    super();
  }

  async suspend() {
    this.samplesGeneratorSubscription &&
      this.samplesGeneratorSubscription.unsubscribe();
  }

  samplesGeneratorSubscription: Subscription | undefined = undefined;

  async resume() {
    // Synthesize samples that get sent to the wasm processor.
    this.samplesGeneratorSubscription = of(animationFrame.now(), animationFrame)
      .pipe(
        repeat(),
        map((start) => animationFrame.now() - start),
        map((t) => this.produceFrame(t))
      )
      .subscribe((frame) => this.tick(frame));
  }

  private sendSamplesToWasm(frame: { t: number; noteFreq: number }) {
    const msElapsed = frame.t - this.prevNoteT;
    const chunks = Math.floor(
      (AudioSynthesizer.UPDATES_PER_SECOND * msElapsed) / 1000
    );
    for (let i = 0; i < chunks; i++) {
      for (let dx = 0; dx < AudioSynthesizer.CHUNK_SIZE; dx++)
        this.wasmChunkBuffer[dx] = this.sinewaveGenerator.next(frame.noteFreq);
      this.wasmSamplesProcessor.add_samples_chunk(this.wasmChunkBuffer);
    }
    this.prevNoteT = frame.t;
  }

  switchTo(
    config: Pick<PitchProducerConfig, "piece" | "startNote" | "bpm">
  ): void {
    this.pitchProducer.send({
      type: "START_PIECE",
      config: {
        ...config,
        startTime: nonNegInteger(this.prevNoteT),
      },
    });
  }

  produceFrame(t: number) {
    const { producer } = this.pitchProducer.state.context;
    if (!producer) {
      throw Error(
        "Retrieving values from the audio synthesizer without first having started a piece."
      );
      // return { t, noteFreq: 0 };
    }

    return {
      t,
      noteFreq: producer(
        nonNegInteger(t)
      ) /*pitchAtMs(t, this.config, this.memo)*/,
    };
  }

  tick(frame: { t: number; noteFreq: number }) {
    this.sendSamplesToWasm(frame);

    if (!this.wasmSamplesProcessor.has_sufficient_samples(this.pitchDetector)) {
      return;
    }

    this.wasmSamplesProcessor.set_latest_samples_on(this.pitchDetector);

    const result = this.pitchDetector.pitches();

    if (result.code !== "success") {
      console.log("error getting pitches", result.message);
    } else {
      const pitches = result.pitches;
      if (pitches.length > 0) {
        pitches.forEach((pitch: Pitch) => {
          if (pitch.onset) {
            this.next({
              type: "onset",
              t: pitch.t,
            });
          } else {
            this.next({
              type: "pitch",
              pitch: {
                clarity: pitch.clarity,
                frequency: pitch.frequency,
                onset: false,
                t: pitch.t,
              },
            });
          }
        });

        pitches.forEach((p) => p.free());
      }
    }
  }

  static async create(
    config: SynthesizerConfig,
    windowSamples = 2048,
    powerThreshold = 0.9,
    clarityThreshold = 0.75
  ) {
    try {
      const wasmBytes = await AudioSynthesizer.fetchMusicAnalyzerWasm();

      await init(WebAssembly.compile(wasmBytes));

      const wasmSamplesProcessor = AudioSamplesProcessor.new();

      const pitchDetector = wasmSamplesProcessor.create_pitch_detector(
        "McLeod",
        windowSamples,
        AudioSynthesizer.SAMPLE_RATE,
        powerThreshold,
        clarityThreshold
      );

      if (!pitchDetector) {
        throw Error("Wasm pitch detector could not be created");
      }

      const result = new AudioSynthesizer(
        config,
        wasmSamplesProcessor,
        pitchDetector
      );

      result.switchTo(config);

      return result;
    } catch (e) {
      console.warn(e);
      throw new Error(
        `Failed to load audio analyzer WASM module. Further info: ${e.message}`
      );
    }
  }

  private static scriptUrl(scriptPath: string) {
    const publicUrl = isTest() ? "./public" : window.location.href;

    return `${publicUrl}/${scriptPath}`;
  }

  private static async fetchMusicAnalyzerWasm() {
    const res = await globalThis.fetch(
      AudioSynthesizer.scriptUrl(
        "music-analyzer-wasm-rs/music_analyzer_wasm_rs_bg.wasm"
      )
    );

    return res.arrayBuffer();
  }
}
