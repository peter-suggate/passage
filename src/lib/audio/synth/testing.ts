import { AudioSynthesizer } from "../recorder/synthaudio/AudioSynthesizer";
import { bpm, posInteger, nonNegInteger } from "@/lib/scales";

export async function majorScaleSynth(BPM = 60, windowSamples = 2048) {
  return await AudioSynthesizer.create(
    {
      bpm: bpm(BPM),
      instrument: "bell",
      scaleType: {
        scale: { tonic: "C", mode: "major" },
        octaves: posInteger(1),
        startOctave: nonNegInteger(5),
      },
    },
    windowSamples,
    0.7,
    0.75
  );
}
