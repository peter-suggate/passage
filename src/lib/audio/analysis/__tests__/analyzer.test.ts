import { allEvents$, expectEvents$ } from "@/lib/testing/rx-testing";
import {
  nearestNotes$,
  recentDistinctNotes$,
  closestMatchingPieces$,
} from "../analyzer";
import { AudioSynthesizer } from "../../recorder/synthaudio/AudioSynthesizer";
import { bpm, nonNegInteger, posInteger, Note } from "@/lib/scales";
import { NearestNote } from "../nearestNote";

async function majorScaleSynth(BPM = 60, windowSamples = 2048) {
  return await AudioSynthesizer.create(
    {
      bpm: bpm(BPM),
      instrument: "bell",
      scaleType: {
        scale: { tonic: "C", mode: "major" },
        octaves: posInteger(1),
        startOctave: nonNegInteger(5),
      },
    },
    windowSamples,
    0.7,
    0.75
  );
}

it("returns all expected notes", async (done) => {
  const NOTES_PER_SECOND = 15;

  const synth = await majorScaleSynth(60 * NOTES_PER_SECOND, 1024);

  expect(
    allEvents$<NearestNote>(
      nearestNotes$(synth),
      (events) => {
        expect(
          events.map(
            (e: NearestNote) => e.value /*({ note: e.value, cents: e.cents })*/
          )
        ).toMatchSnapshot();
      },
      done
    )
  );

  for (let n = 1; n <= NOTES_PER_SECOND; n++) {
    synth.tick(synth.produceFrame((n * 1000) / NOTES_PER_SECOND - 1));
  }

  synth.complete();
});

describe("calculating N last distinct notes", () => {
  it("removes repeated notes", async (done) => {
    const synth = await majorScaleSynth();

    expectEvents$<NearestNote[], Note[]>(
      recentDistinctNotes$(nearestNotes$(synth), 5),
      [["C"], ["C", "D"], ["C", "D", "E"], ["C", "D", "E", "F"]],
      done,
      undefined,
      (e) => e.map((p) => p.value)
    );

    for (let n = 1; n < 5; n++) {
      synth.tick(synth.produceFrame(n * 1000 - 1));
    }
  });

  it("keeps only the last N notes", async (done) => {
    const synth = await majorScaleSynth();

    expectEvents$<NearestNote[], Note[]>(
      recentDistinctNotes$(nearestNotes$(synth), 5),
      [
        ["C"],
        ["C", "D"],
        ["C", "D", "E"],
        ["C", "D", "E", "F"],
        ["C", "D", "E", "F", "G"],
        ["D", "E", "F", "G", "A"],
        ["E", "F", "G", "A", "B"],
        ["F", "G", "A", "B", "C"],
        ["G", "A", "B", "C", "B"],
        ["A", "B", "C", "B", "A"],
        ["B", "C", "B", "A", "G"],
      ],
      done,
      undefined,
      (e) => e.map((p) => p.value)
    );

    for (let n = 1; n < 12; n++) {
      synth.tick(synth.produceFrame(n * 1000 - 1));
    }
  });

  it("removes gliding / passing notes because they interfere with music recognition", async (done) => {
    const NOTES_PER_SECOND = 15;

    const synth = await majorScaleSynth(60 * NOTES_PER_SECOND, 1024);

    expect(
      expectEvents$<NearestNote[], Note[]>(
        recentDistinctNotes$(nearestNotes$(synth), 5),
        [
          ["C"],
          ["C", "D"],
          ["C", "D", "E"],
          ["C", "D", "E", "F"],
          ["C", "D", "E", "F", "G"],
          ["D", "E", "F", "G", "A"],
          ["E", "F", "G", "A", "B"],
          ["F", "G", "A", "B", "C"],
          ["G", "A", "B", "C", "B"],
          ["A", "B", "C", "B", "A"],
          ["B", "C", "B", "A", "G"],
          ["C", "B", "A", "G", "F"],
          ["B", "A", "G", "F", "E"],
          ["A", "G", "F", "E", "D"],
          ["G", "F", "E", "D", "C"],
        ],
        done,
        undefined,
        (e) => e.map((p) => p.value)
      )
    );

    for (let n = 1; n <= NOTES_PER_SECOND; n++) {
      synth.tick(synth.produceFrame((n * 1000) / NOTES_PER_SECOND - 1));
    }
  });
});

describe("detecting closest music piece/scale from recorded notes", () => {
  it("returns the correct best match after 10 notes of a major scale have been played", async (done) => {
    const synth = await majorScaleSynth();

    const NOTES = 10;
    const MIN_NOTES = NOTES;
    const MAX_MATCHES = 1;

    expectEvents$(
      closestMatchingPieces$(
        MIN_NOTES,
        MAX_MATCHES
      )(recentDistinctNotes$(nearestNotes$(synth), NOTES)),
      [[{ piece: { name: "Major Scale" }, distance: 0 }]],
      done
    );

    for (let n = 1; n < NOTES + 1; n++) {
      synth.tick(synth.produceFrame(n * 1000 - 1));
    }
  });
});
