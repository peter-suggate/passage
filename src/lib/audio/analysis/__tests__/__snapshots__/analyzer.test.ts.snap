// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns all expected notes 1`] = `
Array [
  "C",
  "C",
  "C",
  "C",
  "C",
  "C",
  "C",
  "C",
  "C",
  "C",
  "C#",
  "C#",
  "D",
  "D",
  "D",
  "D",
  "D",
  "D",
  "D",
  "D",
  "D",
  "D",
  "D",
  "D#",
  "D#",
  "E",
  "E",
  "E",
  "E",
  "E",
  "E",
  "E",
  "E",
  "E",
  "E",
  "E",
  "F",
  "F",
  "F",
  "F",
  "F",
  "F",
  "F",
  "F",
  "F",
  "F",
  "F",
  "F#",
  "F#",
  "G",
  "G",
  "G",
  "G",
  "G",
  "G",
  "G",
  "G",
  "G",
  "G",
  "G",
  "G#",
  "A",
  "A",
  "A",
  "A",
  "A",
  "A",
  "A",
  "A",
  "A",
  "A",
  "A",
  "A#",
  "A#",
  "B",
  "B",
  "B",
  "B",
  "B",
  "B",
  "B",
  "B",
  "B",
  "B",
  "B",
  "C",
  "C",
  "C",
  "C",
  "C",
  "C",
  "C",
  "C",
  "C",
  "C",
  "C",
  "C",
  "C",
  "B",
  "B",
  "B",
  "B",
  "B",
  "B",
  "B",
  "B",
  "B",
  "B",
  "B",
  "A#",
  "A#",
  "A",
  "A",
  "A",
  "A",
  "A",
  "A",
  "A",
  "A",
  "A",
  "A",
  "G#",
  "G#",
  "G",
  "G",
  "G",
  "G",
  "G",
  "G",
  "G",
  "G",
  "G",
  "G",
  "G",
  "F#",
  "F#",
  "F",
  "F",
  "F",
  "F",
  "F",
  "F",
  "F",
  "F",
  "F",
  "F",
  "F",
  "E",
  "E",
  "E",
  "E",
  "E",
  "E",
  "E",
  "E",
  "E",
  "E",
  "E",
  "D#",
  "D",
  "D",
  "D",
  "D",
  "D",
  "D",
  "D",
  "D",
  "D",
  "D",
  "D",
  "C#",
  "C#",
  "C",
  "C",
  "C",
  "C",
  "C",
  "C",
]
`;
